### DISCUSSION TITLE
FR: Document Why the Snapshot Model Is Better Than the Patch Model

### DISCUSSION THREAD

**Discussion by gavinhoward** [jj-vcs/jj#2950] [General]
**Is your feature request related to a problem? Please describe.**

@martinvonz said on a forum that he believes the snapshot model is better than the patch model, but didn't have time to elaborate.

**Describe the solution you'd like**

As someone who is interested in VCS design and appreciates Jujutsu's design, I want to read @martinvonz's thoughts on the matter.

**Describe alternatives you've considered**

I could write it, but that's circlejerk.

I could cold email @martinvonz, but that would be cold and waste his time when others could benefit too.

**Additional context**

The docs in this repo are great, and such a doc would fit in well.

This is meant to be encouragement, not a demand. Feel free to reject and close.

**Comment by ilyagr** [comment]
I feel like, before we write a doc, we could discuss it, so I converted this to a discussion.

My immediate thoughts:

- `jj` borrows very heavily from a patch-based model of a VCS for its UI. However, you are correct that it is entirely snapshot-based in terms of its storage.

- There are some discussions on Discord about Pijul. My impression is that *purely* patch-based VCS do have some limitations that I don't immediately know how to overcome. For one, it's hard to even print a patch in a way that's easy for the user to interpret, since a patch does not contain the context around the changed lines. Also, in practice patches seem to depend on other patches often, more often than I would like them to as a user, limiting their flexibility. I complain about this [on Discord here](https://discord.com/channels/968932220549103686/968932220549103689/1172661101540941926). I've been wondering whether these things could be improved; I don't know at the moment.

- This is not a clear plan, but I'm vaguely hoping that we find a way to add more patch-based features to `jj`. IMO, there's nothing stopping us from storing some patch-based information in addition to the snapshots, or generating it on the fly (even in Pijul, patches come from diffing the working copy on disk with a snapshot composed of all the patches in its repo). However, it is not easy to figure out how to present that as a coherent UI and a coherent data model (clear correspondence between the snapshot data and the patches).

I'm also curious if people have more thoughts.

**Update:** I focused on the disadvantages of patch-based VCSes above, but I should explicitly say that the problems they are trying to address are also real. I feel like we're running into such problems more and more. For example, jj's support of conflicts is excellent, but there are some obstacles to improving it further that seem related. I might elaborate later, especially if this becomes clearer in my mind (it's closely related to the vague plans/hopes above).

**Update 2**: Elaborating a little on the last paragraph, there is the `A+A-B=A` problem with conflicts, [see Discord again](https://discord.com/channels/968932220549103686/1123461454859882597). This is a simplification that is extremely convenient for sensible snapshots; if such conflicts weren't simplified, this would look bizarre to users, and waste their time. However, the problem is that it makes little sense from patch theory perspective. The simplification causes the composition of conflicted patches to not be associative in jj, which makes reasoning about and keeping track of simplification of conflicts hard.

**Reply by gavinhoward** [reply]
Thank you!

I hate to be picky, but can you quote the stuff from Discord? I can't access it.

**Reply by ilyagr** [reply]
Sorry, it's not trivial to copy-paste, there are a bunch of meandering posts linked to each other. I might try to summarize it later. 

If you are willing to create a Discord account, you can follow the link https://discord.gg/dkmfj3aGQN (from the README) before following the link I gave (https://discord.com/channels/968932220549103686/968932220549103689/1172661101540941926, also https://discord.com/channels/968932220549103686/1093242474144735322/1094422720877445230). I can understand if you don't want to make a Discord account, but many people seem to like it or already use it.

**Reply by gavinhoward** [reply]
Oh, I have a Discord, I just try not to use it. But I'll use it for this.

Hope you don't mind if I make a private, unshared archive of that info.

**Reply by joyously** [reply]
I have a Discord account, but they flagged me as spam (when their login page got in a loop) and I can't log into it. Are there options to view the pages without logging in?

**Reply by ilyagr** [reply]
I wish there was, but I'm not aware of one, sorry. Perhaps create another account?

I wonder whether we could sync the Discord to some other service (matrix? I think there are a few others), but I don't think I'll look into it in the near future.

**Reply by joyously** [reply]
> I'm also curious if people have more thoughts.

The majority of the software at [Wikipedia - VCS](https://en.wikipedia.org/wiki/Comparison_of_version-control_software#Technical_information) is listed as storing changesets, although Darcs is the only one that says "patch". (Pijul is not listed yet.)

In my limited experience, it seems that both patches and snapshots work okay if all you are doing is marching forward, hardly ever looking back.

Both patches and snapshots have the same trouble of dependencies when you try to rearrange the graph or interleave large formatting changes with logic changes and then try to revert something. I saw this with the WordPress repository, which is in SVN, mirrored in Git, and I imported it to Pijul. Large changes were made to add a different editor, followed by bug fixes and updates for the latest version of PHP(large changeset for small changes). Pijul couldn't remove the editor changes due to dependencies, but I wanted to keep the updates for PHP which were made everywhere. Git couldn't do it either because of merge conflicts and wrong parent (or it could be I don't know Git well enough). I didn't even try in SVN.

I think the solution is a hybrid, as suggested (and a few in the Wiki list are) combined with incremental rebase/merge with help from the human when it gets sticky.

**Update**:  I meant to put the link to the [Darcs Model](https://darcs.net/Using/Model) where it talks about primitive patch and high-level patch, and how merges and deletes are done. The following quote reminds me of `jj` rewriting things:
> Darcs handles all of the cases above using the same basic mechanism. In other words:
> - Merging consists in appending to a patch sequence
> - The contents of darcs patches are automatically adjusted to fit their new context.

**Comment by martinvonz** [comment]
I've been very busy lately with our internal alpha release. @gavinhoward, did you find an answer on Discord?

**Reply by gavinhoward** [reply]
I think so?

**Reply by martinvonz** [reply]
Great. I still think it's reasonable to have a document discussing the pros and cons. I'm just not sure when I'll find time to write it.
